### PR TITLE
Move feature gate package from k8s.io/apiserver to k8s.io/component-base

### PR DESCRIPTION
**Pull request by andrewsykim**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
As per discussion in https://github.com/kubernetes/kubernetes/pull/73604/files#r252809685, proposing to move `k8s.io/apiserver/pkg/util/feature` to `k8s.io/component-base`. I'm not set on the package location `k8s.io/component-base/pkg/feature` so happy to change it if there are better ideas. I also think this package might be better renamed to `featuregate` and we can switch the import alias from `utilfeature` to `featuregate` but maybe for another PR?

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/assign @sttts @liggitt @lavalamp @luxas @mtaufen @stewart-yu

### CODE REVIEW

File: cmd/kube-apiserver/app/options/validation.go
Comment (andrewsykim) [inline cmd/kube-apiserver/app/options/validation.go]: Would `featuregate "k8s.io/component-base/pkg/feature"` be better? Didn't change the import alias to reduce # of changes in this PR but I can do this if it makes sense.

File: pkg/apis/core/validation/BUILD
Comment (liggitt) [inline pkg/apis/core/validation/BUILD]: nothing under `k8s.io/apiserver` should be referenced from vendor in build files...

Comment (liggitt) [reply pkg/apis/core/validation/BUILD]: nevermind, fixed in follow up commit

File: cmd/kube-apiserver/app/options/validation.go
Comment (rosti) [reply cmd/kube-apiserver/app/options/validation.go]: Let's keep the changes in this PR to mostly mechanical. You can followup with another tiny PR on that.

**Review by rosti — COMMENTED** [review]
Thank you @andrewsykim !
Looking great so far! Only some minor gofmt changes need addressing. Can you run `./hack/update-gofmt.sh` to get the verify test working?

File: cmd/kube-apiserver/app/options/validation.go
Comment (andrewsykim) [reply cmd/kube-apiserver/app/options/validation.go]: works for me, though I don't think the follow-up will be tiny ;)

**Comment by rosti** [comment]
@andrewsykim can you rebase the PR, please?

**Comment by andrewsykim** [comment]
done!

**Review by luxas — COMMENTED** [review]
For cleanliness, we said we wouldn't have a pkg directory, because then we'll gather a lot of cruft in there ;)
So IMO this should be top-level like the other things

File: staging/src/k8s.io/component-base/pkg/feature/feature_gate.go
Comment (sttts) [inline staging/src/k8s.io/component-base/pkg/feature/feature_gate.go]: this must not move. We have a techinical debt here that we have this global variable at all. Component-base should only have the library code.

Comment (sttts) [reply staging/src/k8s.io/component-base/pkg/feature/feature_gate.go]: Also keeping this here will make the PR much smaller.

Comment (andrewsykim) [reply staging/src/k8s.io/component-base/pkg/feature/feature_gate.go]: Done! I moved everything except for `DefaultMutableFeatureGate` and `DefaulteFeatureGate` to `k8s.io/component-base/featuregate`.

**Comment by sttts** [comment]
/approve
/lgtm

Needs another rebase. And a top-level approval. 

@jbeda @thockin could you approve? We are changing imports throughout the repo.

**Comment by liggitt** [comment]
PR looks good.

Timing-wise, given today is freeze, do we have a sense for how many 1.14-targeting PRs this will cause to need a rebase (I'd guess a lot)? This is similar to the klog change in 1.13 (both in scope and in timing) that got some [attention during the retrospective](https://docs.google.com/document/d/1mKrJm3N4dC4rfOwa5WcvOptmjpteMcNOg1ZSoACKU1g/edit#bookmark=id.up8fcrrcfed2).

**Comment by andrewsykim** [comment]
> Timing-wise, given today is freeze, do we have a sense for how many 1.14-targeting PRs this will cause to need a rebase (I'd guess a lot)? This is similar to the klog change in 1.13 (both in scope and in timing) that got some attention during the retrospective.

Good point, I have no problems revisiting this for v1.15. I'll try my best to keep this PR in a good state for the rest of the day and defer to you & the release leads on whether you want this merged.

**Comment by liggitt** [comment]
merging it first thing in 1.15 would be great from my perspective, thanks

File: staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json
Comment (cheftako) [inline staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json]: Why do we have "k8s.io/component-base/featuregate" in the import path twice?

File: staging/src/k8s.io/apiserver/Godeps/Godeps.json
Comment (cheftako) [inline staging/src/k8s.io/apiserver/Godeps/Godeps.json]: Again imported twice.

**Comment by andrewsykim** [comment]
FYI I'm going to be revisiting this PR soon, please let me know if we don't want this change anymore :)

**Comment by sttts** [comment]
Please go ahead and rebase.

**Review by rosti — APPROVED** [review]
Thanks @andrewsykim !
/approve

**Review by luxas — APPROVED** [review]
LGTM :+1:
nit: anyone have opinions about the name `k8s.io/component-base/featuregate` vs `k8s.io/component-base/features` vs `k8s.io/component-base/feature`
I'm fine as-is, I'm not just sure if the naming ever was discussed

**Comment by neolit123** [comment]
if the package can end up supporting feature mechanics different from feature gates the "feature(s)" naming is a safer long term bet.

**Review by cheftako — COMMENTED** [review]
/lgtm

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: Can we move this to component-base with the rest of the feature gate code? There are several places we use this outside of the apiserver including some controllers and the gce_disks portions of the gce cloud provider. I feel it would be cleaner to keep the feature gate code in one place and minimize the dependencies which are pulled in when accessing features.

**Review by cheftako — COMMENTED** [review]
To take advantage of this in #77374 we would need DefaultFeatureGate also moved to component-base.

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: hmm, see discussion at https://github.com/kubernetes/kubernetes/pull/74021#discussion_r257702497

Comment (andrewsykim) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: There was a comment on this previously https://github.com/kubernetes/kubernetes/pull/74021#discussion_r257702497. The `DefaultFeatureGate` is a global state that currently (sort of?) belongs to the apiserver. I think it may not make much sense to move it into `k8s.io/component-base` since what we add there may not apply to all components in the future. I don't have strong opinions on this though and happy to move it if it makes sense

**Comment by andrewsykim** [comment]
This is ready for review, but will likely need another update after https://github.com/kubernetes/kubernetes/pull/77412

File: staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json]: can you drop the Godeps.json changes from all the commits? these no longer exist

File: staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go]: https://github.com/kubernetes/kubernetes/pull/77412 removed the need to do this for gce_disks' benefit

File: staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json
Comment (andrewsykim) [reply staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json]: yeah will do, sorry I thought `update-vendor.sh` cleaned those up automatically

Comment (andrewsykim) [reply staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json]: I ended up deleting all the Godeps file (stale updates cause this branch was created before go modules) but I noticed apiextensions-apiserver and kube-aggregator still had Godeps files in them in master. Are those special cases or did we just forgot?

File: staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json]: still have some Godeps.json changes hanging around

File: staging/src/k8s.io/apiserver/Godeps/Godeps.json
Comment (liggitt) [inline staging/src/k8s.io/apiserver/Godeps/Godeps.json]: drop Godeps.json

File: staging/src/k8s.io/cloud-provider/Godeps/Godeps.json
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/Godeps/Godeps.json]: drop Godeps.json

File: staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/Godeps/Godeps.json]: those are spurious, I'll open a PR to drop them

**Comment by liggitt** [comment]
there's still godeps.json changes interleaved with the commits that need dropping

**Comment by sttts** [comment]
integration is red

**Comment by logicalhan** [comment]
> integration is red

It's due to this: https://github.com/kubernetes/kubernetes/issues/77573

File: staging/publishing/rules.yaml
Comment (sttts) [inline staging/publishing/rules.yaml]: can we remove apiserver?

File: staging/publishing/import-restrictions.yaml
Comment (sttts) [inline staging/publishing/import-restrictions.yaml]: can we remove apiserver?

**Comment by luxas** [comment]
@andrewsykim this one needs a rebase. Otherwise, after @sttts' comments are addressed, we can merge

**Pull request merged**
